test/e2e/nodeportlocal_test.go

```diff
@@ -0,0 +1,389 @@
+// limitations under the License.
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/k8s"
+)
+
+const (
+	defaultStartPort = 40000
+	defaultEndPort   = 41000
+	updatedStartPort = 42000
+	updatedEndPort   = 43000
+	nplSvcAnnotation = "nodeportlocal.antrea.io/enabled"
+)
+
+func getNPLAnnotation(t *testing.T, data *TestData, r *require.Assertions, testPodName string) (string, *PodIPs) {
+	var nplAnn string
+	var testPodIP *PodIPs
+	var found bool
+	var err error
+	err = wait.Poll(time.Second, defaultTimeout, func() (bool, error) {
+		updatedPod, _ := data.podWaitFor(defaultTimeout, testPodName, testNamespace, func(pod *corev1.Pod) (bool, error) {
+			return pod.Status.Phase == corev1.PodRunning, nil
+		})
+		testPodIP, err = data.podWaitForIPs(defaultTimeout, testPodName, testNamespace)
+		if err != nil {
+			t.Logf("Error when waiting for IP for Pod '%s': %v", testPodName, err)
+			return false, nil
+		}
+		r.Nil(err, "Failed to get Pod")
+		ann := updatedPod.GetAnnotations()
+		t.Logf("Got annotation %v for pod IP %v", ann, testPodIP.ipv4.String())
+		nplAnn, found = ann[k8s.NPLAnnotationKey]
+		return found, nil
+	})
+	r.Nil(err, "Poll for Pod Annotation check failed: %v", err)
+	return nplAnn, testPodIP
+}
+
+func checkForNPLRuleInIPTABLES(t *testing.T, data *TestData, r *require.Assertions, nplAnnotation []k8s.NPLAnnotation, antreaPod, podIP string, present bool) {
+	cmd := []string{"iptables", "-t", "nat", "-S"}
+	t.Logf("Verifying %d annotations for Pod IP: %s", len(nplAnnotation), podIP)
+	err := wait.Poll(time.Second, defaultTimeout, func() (bool, error) {
+		stdout, _, err := data.runCommandFromPod("kube-system", antreaPod, "antrea-agent", cmd)
+		if err != nil {
+			t.Logf("Error while checking rules in IPTABLES: %v", err)
+			// In case of error, retry
+			return false, nil
+		}
+		for i := range nplAnnotation {
+			ruleSpec := []string{
+				"-p", "tcp", "-m", "tcp", "--dport",
+				fmt.Sprint(nplAnnotation[i].NodePort), "-j", "DNAT", "--to-destination", fmt.Sprintf("%s:%d", podIP, nplAnnotation[i].PodPort),
+			}
+			rs := strings.Join(append([]string{"-A", "ANTREA-NODE-PORT-LOCAL"}, ruleSpec...), " ")
+			t.Logf("Searching for IPTABLES rule: %v", rs)
+
+			if strings.Contains(stdout, rs) && present {
+				t.Logf("Found rule in IPTABLES")
+			} else if !strings.Contains(stdout, rs) && !present {
+				t.Logf("Rule not found in IPTABLES")
+			} else {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	r.Nil(err, "Poll for IPTABLES rules check failed")
+}
+
+func checkTrafficForNPL(t *testing.T, data *TestData, r *require.Assertions, nplAnnotation []k8s.NPLAnnotation, clientName string) {
+	err := wait.Poll(time.Second, defaultTimeout, func() (bool, error) {
```

Comment:
what is the rationale for using `Poll` here? At this stage, if we have verified the Pod annotation and verified that the iptables rules is present, I don't think we should poll / sleep

Reply:
Make sense, we don't nee poll here.